Add method foreachAsync to Sink object

gabfssilva opened the pull request:
I always need to create a Sink that consumes something asynchronously, so, since it is not possible (nor adequate) to use Sink.foreach(future), what about a Sink.foreachAsync(parallelism)(future)?

ps.: I know I need to write tests, but, since I did not found any tests validating unitarily Sink.foreach I'm wondering where I could add the new tests.

gabfssilva committed:
Add method foreachAsync to Sink object

lightbend-cla-validator commented:
Hi @gabfssilva,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

ktoso commented:
I'd be okey with this addition.
Let's see if rest of the team agrees... One would also have to add tests and add documentation for it.

ktoso commented:
OK TO TEST

ktoso commented:
Generally though first open a ticket and then PRs, as requested in the [CONTRIBUTING.md](https://github.com/akka/akka/blob/master/CONTRIBUTING.md) guidelines.

ktoso commented:
Hm... do note that we actually have `Sink.foreachParallel` already:

```
  /**
   * A `Sink` that will invoke the given function to each of the elements
   * as they pass in. The sink is materialized into a [[scala.concurrent.Future]]
   *
   * If `f` throws an exception and the supervision decision is
   * [[akka.stream.Supervision.Stop]] the `Future` will be completed with failure.
   *
   * If `f` throws an exception and the supervision decision is
   * [[akka.stream.Supervision.Resume]] or [[akka.stream.Supervision.Restart]] the
   * element is dropped and the stream continues.
   *
   * See also [[Flow.mapAsyncUnordered]]
   */
  def foreachParallel[T](parallelism: Int)(f: T ⇒ Unit)(implicit ec: ExecutionContext): Sink[T, Future[Done]] =
    Flow[T].mapAsyncUnordered(parallelism)(t ⇒ Future(f(t))).toMat(Sink.ignore)(Keep.right)

```

And yeah it has tests: SinkForeachParallelSpec

gabfssilva commented:
`Sink.foreachParallel` is somehow different. It expects `T => Unit` instead of a `T => Future[Unit]`.

lightbend-cla-validator commented:
Hi @gabfssilva,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala:
```diff
@@ -258,6 +258,15 @@ object Sink {
+   * the stream.
+   */
+  def foreachAsync[T](parallelism: Int)(f: T ⇒ Future[Unit]): Sink[T, Future[Done]] =
+    Flow[T].mapAsync(parallelism)(f).toMat(Sink.ignore)(Keep.right).named("foreachAsyncSink")
```
this should be `mapAsyncUnordered` since the order of the returned Unit values don't matter

gabfssilva replied:
What about the sink processing? Depending, one can be worried about the order. Maybe also create a `foreachAsyncUnordered`?

patriknw replied:
the `f` will be called in order, so I don't see the problem

patriknw commented on akka-docs/src/main/paradox/stream/operators/Sink/foreachAsync.md:
```diff
@@ -0,0 +1,33 @@
+# foreachAsync
```
I'm skeptical to if this adds much value over using `mapAsyncUnorderd` + `Sink.ignore`.
On the other hand we already have `foreachParallel`, which would have been better as `f: T => Future[Unit]` and we don't use the naming `Parallel` in other places.

Therefore it might be good to add this and deprecate `foreachParallel`.

gabfssilva replied:
I think foreachParallel is for different purposes, although we can achieve the same `foreachParallel` behaviour as `f: T => Future[Unit]`  just by using the default dispatcher.

patriknw replied:
if one need the same thing as `foreachParallel` one would use `foreachAsync(Future(f))`

ktoso replied:
Have we reached a decision here?
I'm okey with keeping both in this PR. 
Potentially one could deprecate the Parallel version.

ktoso replied:
I see it's deprecated now, all good then

patriknw replied:
yes, I think we should be careful to not just add more and more to the API. Here it makes sense to trade one for the other.

`foreachAsync` is more consistent with other naming, and it's more powerful. `foreachParallel` can easily be expressed as `foreachAsync` by letting the user decide to spawn a Future or call out to something else.

lightbend-cla-validator commented:
Hi @gabfssilva,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

gabfssilva commented:
@patriknw, what about now? Is there anything else to do when we deprecate something?

Also, I just signed the CLA.

johanandren reviewed (approved):
LGTM

johanandren commented:
Verified that CLA is in database, not sure what cla validator is smoking.

gabfssilva commented:
@johanandren maybe I need to send another commit and then it'll re-check?

ktoso commented:
There's no need for that, manually confirming is fine. 
I'll review now and we can merge if all good

patriknw commented on akka-docs/src/main/paradox/stream/operators/Sink/foreachAsync.md:
```diff
@@ -0,0 +1,33 @@
+
+## Description
+
+Invoke a given procedure asynchronously for each element received. Note that it is not safe to mutate shared state from the procedure.
```
I'm not sure about the wording here. The purpose of `foreach` is to do side effects and that typically involves mutating state somewhere. Perhaps: "Note that if shared state is mutated from the procedure that must be done in a thread-safe way."

patriknw commented on akka-docs/src/main/paradox/stream/operators/Sink/foreachAsync.md:
```diff
@@ -0,0 +1,33 @@
+The sink materializes into a  @scala[`Future[Option[Done]]`] @java[`CompletionStage<Optional<Done>`] which completes when the
+stream completes, or fails if the stream fails.
+
+Note that it is not safe to mutate state from the procedure.
```
repeating this one more time is not needed.

patriknw commented on akka-docs/src/main/paradox/stream/operators/Sink/foreachAsync.md:
```diff
@@ -0,0 +1,33 @@
+
+@ref[Sink stages](../index.md#sink-stages)
+
+@@@div { .group-scala }
```
no java signature?

gabfssilva replied:
there is no "@scala-signature" nor "@java-signature", just "@signature" which translates to Scala code, so, I can't see the point to add it. is there any other way to add the Java signature?

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:
```diff
@@ -109,7 +120,9 @@ object Sink {
    * element is dropped and the stream continues.
+   * @deprecated Use `foreachAsync` instead, it allows you to choose your own execution context.
    */
+  @deprecated("Use `foreachAsync` instead, it allows you to choose your own execution context.", since = "2.5.13")
```
"it allows you to choose your own execution context"

well, that was the case for `foreachParallel` also, since it's a parameter here

"it allows you to choose how to run the procedure, by calling some other API returning a CompletionStage or using CompletableFuture.supplyAsync"

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala:
```diff
@@ -290,7 +301,9 @@ object Sink {
    * See also [[Flow.mapAsyncUnordered]]
+   * @deprecated Use `foreachAsync` instead, it allows you to choose your own execution context.
    */
+  @deprecated("Use `foreachAsync` instead, it allows you to choose your own execution context.", since = "2.5.13")
```
"by calling some other API returning a Future or spawning a new Future"

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala:
```diff
@@ -290,7 +301,9 @@ object Sink {
    * element is dropped and the stream continues.
    *
    * See also [[Flow.mapAsyncUnordered]]
+   * @deprecated Use `foreachAsync` instead, it allows you to choose your own execution context.
```
I don't think that is needed, the Scala `@deprecated` annotation should be enough

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:
```diff
@@ -109,7 +120,9 @@ object Sink {
    * If `f` throws an exception and the supervision decision is
    * [[akka.stream.Supervision.Resume]] or [[akka.stream.Supervision.Restart]] the
    * element is dropped and the stream continues.
+   * @deprecated Use `foreachAsync` instead, it allows you to choose your own execution context.
```
I don't think that is needed, the Scala @deprecated annotation should be enough

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:
```diff
@@ -99,6 +99,17 @@ object Sink {
+   * normal end of the stream, or completed with `Failure` if there is a failure signaled in
+   * the stream.
+   *
+   * See also [[Flow.mapAsyncUnordered]].
```
not sure if this reference is needed, might just be confusing?

patriknw commented on akka-docs/src/main/paradox/stream/operators/Sink/foreachAsync.md:
```diff
@@ -0,0 +1,33 @@
+
+Note that it is not safe to mutate state from the procedure.
+
+
```
We are trying to improve the quality and detail in the documentation of the operators so this one should probably include code examples of the 2 cases:
1. calling another API returning a Future/CompletionStage
1. spawning a Future/CompletableFuture, the latter is done with `CompletableFuture.supplyAsync`

patriknw reviewed (commented):
some more quality/documentation improvements...

gabfssilva commented:
Okay, so:

- I removed the `@deprecated` in the comments and also the `see also Flow.mapAsyncUnordered`.
- Added code snips for both Java and Scala docs.
- Changed the Java api signature to `f: Function[T, Void]` instead of `f: Function[T, Unit]`.
- Document fixing: removing duplicated comments, added the suggestion from @patriknw and corrected the sink materialization return from `Future[Option[Done]]` to `Future[Done]` and `CompletionStage<Optional<Done>>` to CompletionStage<Done> in both Scala and Java docs, respectively.

johanandren commented:
Tried to reproduce locally as well but it passes as far as I can see. The logic that fails is verifying that the link back to the category is one of the known categories.

You think you could sneak some more info into the `StreamOperatorsIndexGenerator` output, for example include `categories` in the failure message in the `require` on line 229 so we can see if there is something surprising going on there?

gabfssilva commented:
Tried to "println" the categories.

Jenkins prints:

```
List(Source operators, Sink operators, Additional Sink and Source converters, File IO Sinks and Sources, Simple operators, Flow operators composed of Sinks and Sources, Asynchronous operators, Timer driven operators, Backpressure aware operators, Nesting and flattening operators, Time aware operators, Fan-in operators, Watching status operators, Actor interop operators)
```

And locally:

```
List(Source stages, Sink stages, Additional Sink and Source converters, File IO Sinks and Sources, Simple operators, Flow stages composed of Sinks and Sources, Asynchronous operators, Timer driven stages, Backpressure aware stages, Nesting and flattening stages, Time aware stages, Fan-in stages, Watching status stages, Actor interop stages)
```

No idea why this happening, but, I think it has happened after I added the code snips.

Anyone?

gabfssilva commented:
It seems like the `categories` had changed and since I did not merge with master upstream it couldn't work on Jenkins. Just have merged, working as expected again. =P

gabfssilva commented:
is there anything else to do over here?

johanandren reviewed (approved):
LGTM, sorry for the delayed review.

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkForeachAsyncSpec.scala:
```diff
@@ -0,0 +1,128 @@
+      def one = { oneCalled = true; 1 }
+      def two = { twoCalled = true; 2 }
+      def three = { threeCalled = true; 3 }
+      def four = { fourCalled = true; 4 }
```
this is not completely thread safe, would need `@volatile` on the variables

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkForeachAsyncSpec.scala:
```diff
@@ -0,0 +1,128 @@
+      val sink: Sink[() ⇒ Int, Future[Done]] = {
+        Sink.foreachAsync(1) { (n: () ⇒ Int) ⇒
+          Future {
+            latch(n())
```
perhaps you intended to do `Await` on the latch?

is the sleep needed then?

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkForeachAsyncSpec.scala:
```diff
@@ -0,0 +1,128 @@
+      val sink: Sink[Int, Future[Done]] = {
+        Sink.foreachAsync(4) { n: Int ⇒
+          Future {
+            latch(n)
```
await?

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkForeachAsyncSpec.scala:
```diff
@@ -0,0 +1,128 @@
+    "handle empty source" in assertAllStagesStopped {
+      import system.dispatcher
+      val p = Source(List.empty[Int]).runWith(Sink.foreachAsync(3)(a ⇒ Future {}))
+      Await.result(p, 200.seconds)
```
`remainingOrDefault` would be enough instead of 200

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkForeachAsyncSpec.scala:
```diff
@@ -0,0 +1,128 @@
+      assert(fourCalled)
+    }
+  }
+
```
Could you also take a look at the existing `SinkForeachParallelSpec` and see if some of it can be incorporated here, but implemented with `foreachAsync`.

patriknw reviewed (commented):
Looking good, a few more things around the test...

johanandren commented:
`assertAllStagesStopped` was moved in PR #25234 so you'll have to update the tests

gabfssilva commented:
Today I'll correct everything and take a look at @patriknw tips. Sorry for the delay, not having much time lately. =/

gabfssilva commented:
So:

- added all the SinkForeachParallelSpec tests to SinkForeachAsyncSpec;
- @volatile added to the "back-pressure upstream elements when downstream is slow" test; and
- waiting for the result on `latch(n)`

johanandren reviewed (approved):
LGTM

gabfssilva commented on akka-docs/src/test/java/jdocs/stream/SinkRecipeDocTest.java:
```diff
@@ -0,0 +1,45 @@
+import jdocs.AbstractJavaTest;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import scala.runtime.BoxedUnit;
```
removed.

johanandren commented:
Sorry about the slow pace of review on this one. Vacation times so we are spread thin in the team.

gabfssilva commented:
No problem. I'm trying to keep this PR updated so it can be easy to merge when you guys are ready. =)

patriknw commented on akka-docs/src/main/paradox/stream/operators/Sink/foreachAsync.md:
```diff
@@ -0,0 +1,38 @@
+
+@@@div { .callout }
+
+**cancels** never
```
not even when the procedure throws or if the Future is completed with exception?

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:
```diff
@@ -99,6 +99,15 @@ object Sink {
+   * normal end of the stream, or completed with `Failure` if there is a failure signaled in
+   * the stream.
+   */
+  def foreachAsync[T](parallelism: Int)(f: function.Function[T, CompletionStage[Void]]): Sink[T, CompletionStage[Done]] =
```
We haven't used `CompletionStage[Void]` anywhere else. Wouldn't it be more consistent to use `CompletionStage[Done]`?

gabfssilva replied:
Actually I don't think it applies in this case, I'll tell why.
If we use CompletionStage[Done]:

```java
final CompletableFuture<Done> doneCompletableFuture = CompletableFuture
  .completedFuture("a")
  .thenAccept(v -> {
      //accept the data
  })
  .thenApply(v -> Done.done());
```
Don't know about you, but, this last line just to map to Done annoys me a lot. =P haha

If we use Void:

```java
final CompletableFuture<Void> voidCompletableFuture = CompletableFuture
  .completedFuture("a")
  .thenAccept(v -> {
    //accept the data
  });
```
No need to do unnecessary maps. =)

Also, I'm using Future[Unit] in the scaladsl, so, semantically I guess it'd be better to use Void in the javadsl.

But, if you have another point I'd be glad to hear.

jypma replied:
Wasn't `Done` also introduced because `Void` doesn't have an instance (you can only pass `null`), and `null` isn't allowed inside the reactive streams spec (nor as materialized values)?

So introducing `Void` in the API would limit the ability for that part to become a stream or materialized value later on. Even though a `Source[Done]` would be rather silly :-)

gabfssilva replied:
@patriknw in this case you wouldn't use `CompletableFuture.supplyAsync`, rather you should use `CompletableFuture.runAsync`:

```java
final Sink<Integer, CompletionStage<Done>> sinkCompletion = 
    Sink.foreachAsync(10, param -> CompletableFuture.runAsync(System.out::println));
```

patriknw replied:
I’m convinced. We should see this as foreach.

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:
```diff
@@ -110,6 +119,7 @@ object Sink {
    * [[akka.stream.Supervision.Resume]] or [[akka.stream.Supervision.Restart]] the
    * element is dropped and the stream continues.
    */
+  @deprecated("Use `foreachAsync` instead, it allows you to choose how to run the procedure, by calling some other API returning a CompletionStage or using CompletableFuture.supplyAsync.", since = "2.5.13")
```
since 2.5.17

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala:
```diff
@@ -310,6 +319,7 @@ object Sink {
    *
    * See also [[Flow.mapAsyncUnordered]]
    */
+  @deprecated("Use `foreachAsync` instead, it allows you to choose how to run the procedure, by calling some other API returning a Future or spawning a new Future.", since = "2.5.13")
```
and here

patriknw reviewed (commented):
sorry for delay, let's get this in now.
some minor nitpicks though...

patriknw commented on akka-docs/src/main/paradox/stream/operators/Sink/foreachAsync.md:
```diff
@@ -0,0 +1,38 @@
+
+Java
+:   @@snip [SinkRecipeDocTest.java]($code$/java/jdocs/stream/SinkRecipeDocTest.java) { #forseachAsync-processing }
+
```
we have changed snippet paths in master, please rebase and change to
```
Scala
:   @@snip [SinkRecipeDocSpec.scala](/akka-docs/src/test/scala/docs/stream/SinkRecipeDocSpec.scala) { #forseachAsync-processing }

Java
:   @@snip [SinkRecipeDocTest.java](/akka-docs/src/test/java/jdocs/stream/SinkRecipeDocTest.java) { #forseachAsync-processing }
```

patriknw commented on akka-docs/src/main/paradox/stream/operators/Sink/foreachAsync.md:
```diff
@@ -0,0 +1,38 @@
+
+## Signature
+
+@@signature [Sink.scala]($akka$/akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala) { #foreachAsync }
```
we have changed snippet paths in master, please rebase and change to

```
@@signature [Sink.scala](/akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala) { #foreachAsync }
```

patriknw reviewed (commented):
just one more thing, then I will approve and merge...

patriknw reviewed (approved):
LGTM

patriknw commented:
Thanks @gabfssilva

patriknw commented:
Refs https://github.com/akka/akka/issues/25622
